### PR TITLE
[linker] Do not mark NSObject subclasses from Xamarin.Forms.Platform.iOS.dll assembly

### PR DESCRIPTION
**Pull request by spouliot**
Turn older #7165 prototype into an experimental feature. It can be
enabled by adding `--optimize=experimental-xforms-product-type` to the
**Additional mtouch arguments** of the project.

ref: https://github.com/xamarin/xamarin-macios/pull/7165

### CODE REVIEW

File: tools/linker/MarkNSObjects.cs [hunk 1]
```diff
@@ -151,9 +152,15 @@ static bool IsProductMethod (MethodDefinition method)
...
+		bool IsProductType (TypeDefinition type)
 		{
-			return type.Module.Assembly.Name.Name == ProductAssembly;
+			var name = type.Module.Assembly.Name.Name;
+			switch (name) {
+			case "Xamarin.Forms.Platform.iOS":
+				return LinkContext.App.Optimizations.ExperimentalFormsProductType == true;
+			default:
+				return name == ProductAssembly;
+			}
```
Comment (mandel-macaque) [inline tools/linker/MarkNSObjects.cs]: Oh! You misses the chance to use the new C# switch expressions!

Comment (spouliot) [reply tools/linker/MarkNSObjects.cs]: I have not yet looked at the IL yet - so I'm sticking to known patterns ;-)

**Comment by spouliot** [comment]
16x HE0038 (watchOS deploy issue)

32bits iOS timeouts
* some report HE0051 -> https://github.com/xamarin/xamarin-macios/pull/7473
* others just timeout without -> https://github.com/xamarin/maccore/issues/2069

mmptest failure -> to be fixed (need to update tool message to include new optimization)
MTouch tests -> same ^

**Comment by spouliot** [comment]
failure is a known issue, HE0038 on watchOS -> https://github.com/xamarin/maccore/issues/581

**Comment by spouliot** [comment]
@monojenkins backport to xcode11.3

**Pull request merged**